File: lib/oxide/expression.rb
require('oxide/values/boolean')
require('oxide/values/floating_point')
require('oxide/values/function')
require('oxide/values/integer')

# rubocop:disable Metrics/ClassLength
module Oxide
  class Expression
    class << self
      def evaluate(attributes, context)
        new(attributes).evaluate(context)
      end
    end

    attr_reader :attributes

    def initialize(attributes)
      @attributes = attributes
    end

    # rubocop:disable Metrics/AbcSize
    # rubocop:disable Metrics/CyclomaticComplexity
    # rubocop:disable Metrics/MethodLength
    def evaluate(context)
      case type
      when 'or'
        evaluate_left(context).or(evaluate_right(context))
      when 'and'
        evaluate_left(context).and(evaluate_right(context))
      when 'equal'
        evaluate_left(context) == evaluate_right(context)
      when 'not_equal'
        evaluate_left(context) != evaluate_right(context)
      when 'less_than'
        evaluate_left(context) < evaluate_right(context)
      when 'less_than_or_equal'
        evaluate_left(context) <= evaluate_right(context)
      when 'greater_than'
        evaluate_left(context) > evaluate_right(context)
      when 'greater_than_or_equal'
        evaluate_left(context) >= evaluate_right(context)
      when 'pipe'
        evaluate_right(context).evaluate(evaluate_left(context))
      when 'application'
        evaluate_left(context).evaluate(evaluate_right(context))
      when 'addition'
        evaluate_left(context) + evaluate_right(context)
      when 'subtraction'
        evaluate_left(context) - evaluate_right(context)
      when 'multiplication'
        evaluate_left(context) * evaluate_right(context)
      when 'division'
        evaluate_left(context) / evaluate_right(context)
      when 'modulo'
        evaluate_left(context) % evaluate_right(context)
      when 'expressions'
        evaluate_expressions(context).last
      when 'false'
        Oxide::Values::Boolean.new(false)
      when 'true'
        Oxide::Values::Boolean.new(true)
      when 'integer'
        Oxide::Values::Integer.new(integer)
      when 'floating_point'
        Oxide::Values::FloatingPoint.new(floating_point)
      when 'identifier'
        context.fetch(identifier)
      when 'assignment'
        context[identifier] = evaluate_right(context)
      when 'addition_assignment'
        context[identifier] =
          context.fetch(identifier) + evaluate_right(context)
      when 'subtraction_assignment'
        context[identifier] =
          context.fetch(identifier) - evaluate_right(context)
      when 'multiplication_assignment'
        context[identifier] =
          context.fetch(identifier) * evaluate_right(context)
      when 'division_assignment'
        context[identifier] =
          context.fetch(identifier) / evaluate_right(context)
      when 'modulo_assignment'
        context[identifier] =
          context.fetch(identifier) % evaluate_right(context)
      when 'function'
        Oxide::Values::Function.new(identifier, right, context)
      else
        raise(ArgumentError, "invalid expression type: #{type.to_json}")
      end
    end
    # rubocop:enable Metrics/AbcSize
    # rubocop:enable Metrics/CyclomaticComplexity
    # rubocop:enable Metrics/MethodLength

    def evaluate_left(context)
      Oxide::Expression.evaluate(left, context)
    end

    def evaluate_right(context)
      Oxide::Expression.evaluate(right, context)
    end

    def evaluate_expressions(context)
      expressions.map do |expression|
        Oxide::Expression.evaluate(expression, context)
      end
    end

    private

    def type
      attributes.fetch('type')
    end

    def left
      attributes.fetch('left')
    end

    def right
      attributes.fetch('right')
    end

    def expressions
      attributes.fetch('expressions')
    end

    def integer
      attributes.fetch('integer')
    end

    def floating_point
      attributes.fetch('floating_point')
    end

    def identifier
      attributes.fetch('identifier')
    end
  end
end
# rubocop:enable Metrics/ClassLength

File: ext/parser/expression.c
#include <ruby.h>

#include "expression.h"
#include "expressions.h"

Expression *create_expression(ExpressionType type, long integer, double floating_point, char *identifier, Expression *expression, Expression *left, Expression *right, Expressions *expressions) {
  Expression *this = (Expression *)malloc(sizeof(Expression));

  if(this == NULL)
    return NULL;

  this->type = type;
  this->integer = integer;
  this->floating_point = floating_point;
  this->identifier = identifier;
  this->expression = expression;
  this->left = left;
  this->right = right;
  this->expressions = expressions;

  return this;
}

VALUE expression_to_hash(Expression *this) {
  if(this == NULL)
    return Qnil;

  VALUE rb_vexpression = rb_hash_new();

  switch(this->type) {
    case FALSE_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("false")));
      break;
    case TRUE_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("true")));
      break;
    case INTEGER_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("integer")));
      break;
    case FLOATING_POINT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("floating_point")));
      break;
    case IDENTIFIER_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("identifier")));
      break;
    case ASSIGNMENT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("assignment")));
      break;
    case ADDITION_ASSIGNMENT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("addition_assignment")));
      break;
    case SUBTRACTION_ASSIGNMENT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("subraction_assignment")));
      break;
    case MULTIPLICATION_ASSIGNMENT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("multiplication_assignment")));
      break;
    case DIVISION_ASSIGNMENT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("division_assignment")));
      break;
    case MODULO_ASSIGNMENT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("modulo_assignment")));
      break;
    case MAPPED_TO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("mapped_to")));
      break;
    case LOGICAL_OR_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("logical_or")));
      break;
    case LOGICAL_AND_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("logical_and")));
      break;
    case EQUAL_TO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("equal_to")));
      break;
    case NOT_EQUAL_TO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("not_equal_to")));
      break;
    case LESS_THAN_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("less_than")));
      break;
    case LESS_THAN_OR_EQUAL_TO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("less_than_or_equal_to")));
      break;
    case GREATER_THAN_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("greater_than")));
      break;
    case GREATER_THAN_OR_EQUAL_TO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("greater_than_or_equal_to")));
      break;
    case PIPE_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("pipe")));
      break;
    case APPLICATION_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("application")));
      break;
    case ADDITION_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("addition")));
      break;
    case SUBTRACTION_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("subtraction")));
      break;
    case MULTIPLICATION_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("multiplication")));
      break;
    case DIVISION_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("division")));
      break;
    case MODULO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("modulo")));
      break;
    case EXPRESSIONS_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("type")), ID2SYM(rb_intern("expressions")));
      break;
  }

  switch(this->type) {
    case INTEGER_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("integer")), INT2NUM(this->integer));
      break;
    case FLOATING_POINT_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("floating_point")), rb_float_new(this->floating_point));
      break;
    case IDENTIFIER_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("identifier")), rb_str_new2(this->identifier));
      break;
    case ASSIGNMENT_EXPRESSION:
    case ADDITION_ASSIGNMENT_EXPRESSION:
    case SUBTRACTION_ASSIGNMENT_EXPRESSION:
    case MULTIPLICATION_ASSIGNMENT_EXPRESSION:
    case DIVISION_ASSIGNMENT_EXPRESSION:
    case MODULO_ASSIGNMENT_EXPRESSION:
    case MAPPED_TO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("identifier")), rb_str_new2(this->identifier));
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("expression")), expression_to_hash(this->expression));
      break;
    case LOGICAL_OR_EXPRESSION:
    case LOGICAL_AND_EXPRESSION:
    case EQUAL_TO_EXPRESSION:
    case NOT_EQUAL_TO_EXPRESSION:
    case LESS_THAN_EXPRESSION:
    case LESS_THAN_OR_EQUAL_TO_EXPRESSION:
    case GREATER_THAN_EXPRESSION:
    case GREATER_THAN_OR_EQUAL_TO_EXPRESSION:
    case PIPE_EXPRESSION:
    case APPLICATION_EXPRESSION:
    case ADDITION_EXPRESSION:
    case SUBTRACTION_EXPRESSION:
    case MULTIPLICATION_EXPRESSION:
    case DIVISION_EXPRESSION:
    case MODULO_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("left")), expression_to_hash(this->left));
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("right")), expression_to_hash(this->right));
      break;
    case EXPRESSIONS_EXPRESSION:
      rb_funcall(rb_vexpression, rb_intern("[]="), 2, ID2SYM(rb_intern("expressions")), expressions_to_hash(this->expressions));
      break;
    default:
      break;
  }

  return rb_vexpression;
}

void delete_expression(Expression *this) {
  if(this == NULL)
    return;

  if(this->identifier != NULL)
    free(this->identifier);

  if(this->expression != NULL)
    delete_expression(this->expression);

  if(this->left != NULL)
    delete_expression(this->left);

  if(this->right != NULL)
    delete_expression(this->right);

  if(this->expressions != NULL)
    delete_expressions(this->expressions);

  free(this);
}
